Usage of non-exposed pcap function

**Elasticsearch version**: irrelevant

**Moloch version**: 2.1

**OS name and version**: 5.3.8-1-MANJARO

**How was Moloch built/installed:** (rpm, deb, easybutton, ...): modified install for arch based systems: `./configure --with-libpcap=no --with-yara=no --with-glib2=no --with-pfring=no --with-curl=no --with-lua=no LIBS="-lpcap -lyara -llua -lcurl" GLIB2_CFLAGS="-I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include" GLIB2_LIBS="-lglib-2.0 -lgmodule-2.0 -lgobject-2.0 -lgio-2.0`

**Provide logs, stack traces and steps to reproduce**: (inside easybutton --install, while building capture)
``` 
gcc -rdynamic -ggdb -pthread main.o db.o yara.o http.o config.o parsers.o plugins.o field.o trie.o writers.o writer-inplace.o writer-disk.o writer-null.o writer-simple.o readers.o reader-libpcap-file.o reader-libpcap.o reader-tpacketv3.o reader-null.o packet.o session.o rules.o drophash.o pq.o -o moloch-capture \
            -u g_checksum_update -u g_hmac_update -u g_uri_unescape_segment \
    -lnl-genl-3 -lnl-3 \
    -lglib-2.0 -lgmodule-2.0 -lgobject-2.0 -lgio-2.0 -lmaxminddb -lnet -lrt -luuid -lpcre -lpcap -lyara -llua -lcurl thirdparty/http_parser.o thirdparty/js0n.o thirdparty/patricia.o -ldl -lssl -lcrypto -lyaml \
    -lm -lresolv -lmagic -lffi -lz
/usr/bin/ld: reader-libpcap-file.o: in function `reader_libpcapfile_opened':
/home/m/IdeaProjects/moloch/capture/reader-libpcap-file.c:562: undefined reference to `dlt_to_linktype'
/usr/bin/ld: reader-libpcap.o: in function `reader_libpcap_start':
/home/m/IdeaProjects/moloch/capture/reader-libpcap.c:97: undefined reference to `dlt_to_linktype'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:41: all] Error 1
make[1]: Leaving directory '/home/m/IdeaProjects/moloch/capture'
make: *** [Makefile:329: all-recursive] Error 1
MOLOCH: make failed
```

**Problem:** The used function is only exposed in the (statically linked) .a version of the pcap lib, but not in the .so, as it's not (yet) exposed in the pcap lib, or in any pcap.h. You can test this with `objdump -t libpcap.a | grep dlt` vs `objdump -T libpcap.so | grep dlt`

This PR over there fixes it: https://github.com/the-tcpdump-group/libpcap/pull/882, in the meantime a copy of the fixed libpcap.so is hosted at https://öä.eu/libpcap.so

This issue is a pre-requisite for my efforts to make it arch-linux installable / non-statically-linked installable on systems whose package managers support the according dependencies.

--- a/capture/main.c
+++ b/capture/main.c
@@ -653,8 +653,8 @@ gboolean moloch_ready_gfunc (gpointer UNUSED(user_data))
         }
     }
     moloch_reader_start();
-    if (!config.pcapReadOffline && (pcapFileHeader.linktype == 0 || pcapFileHeader.snaplen == 0))
-        LOGEXIT("Reader didn't call moloch_packet_set_linksnap");
+    if (!config.pcapReadOffline && (pcapFileHeader.dlt == DLT_NULL || pcapFileHeader.snaplen == 0))
+        LOGEXIT("Reader didn't call moloch_packet_set_dltsnap");
     return FALSE;
 }
 /******************************************************************************/
@@ -737,7 +737,7 @@ LLVMFuzzerInitialize(int *UNUSED(argc), char ***UNUSED(argv))
     config.nodeName = strdup("fuzz");
 
     hashSalt = 0;
-    pcapFileHeader.linktype = 1;
+    pcapFileHeader.dlt = DLT_EN10MB;
 
     moloch_free_later_init();
     moloch_hex_init();

--- a/capture/moloch.h
+++ b/capture/moloch.h
@@ -706,7 +706,7 @@ typedef struct {
 	int32_t  thiszone;	/* gmt to local correction */
 	uint32_t sigfigs;	/* accuracy of timestamps */
 	uint32_t snaplen;	/* max length saved portion of each pkt */
-	uint32_t linktype;	/* data link type (LINKTYPE_*) */
+	uint32_t dlt;	        /* data link type - see https://github.com/aol/moloch/issues/1303#issuecomment-554684749 */
 } MolochPcapFileHdr_t;
 
 #ifndef likely
@@ -1039,7 +1039,9 @@ void     moloch_packet_batch_flush(MolochPacketBatch_t *batch);
 void     moloch_packet_batch(MolochPacketBatch_t * batch, MolochPacket_t * const packet);
 void     moloch_packet_batch_process(MolochPacketBatch_t * batch, MolochPacket_t * const packet, int thread);
 
-void     moloch_packet_set_linksnap(int linktype, int snaplen);
+void     moloch_packet_set_dltsnap(int dlt, int snaplen);
+void     moloch_packet_set_linksnap(int linktype, int snaplen); // Don't use, backwards compat
+uint32_t moloch_packet_dlt_to_linktype(int dlt);
 void     moloch_packet_drophash_add(MolochSession_t *session, int which, int min);
 
 void     moloch_packet_save_ethernet(MolochPacket_t * const packet, uint16_t type);

--- a/capture/packet.c
+++ b/capture/packet.c
@@ -21,6 +21,7 @@
 #include <arpa/inet.h>
 #include <net/ethernet.h>
 #include <errno.h>
+#include "pcap.h"
 
 //#define DEBUG_PACKET
 
@@ -303,7 +304,7 @@ LOCAL void moloch_packet_process(MolochPacket_t *packet, int thread)
     if (session->packets[packet->direction] <= 10) {
         const uint8_t *pcapData = packet->pkt;
 
-        if (pcapFileHeader.linktype == 1) {
+        if (pcapFileHeader.dlt == DLT_EN10MB) {
             if (packet->direction == 1) {
                 moloch_field_macoui_add(session, mac1Field, oui1Field, pcapData+0);
                 moloch_field_macoui_add(session, mac2Field, oui2Field, pcapData+6);
@@ -1189,12 +1190,12 @@ void moloch_packet_batch(MolochPacketBatch_t * batch, MolochPacket_t * const pac
     MolochPacketRC rc;
 
 #ifdef DEBUG_PACKET
-    LOG("enter %p %u %d", packet, pcapFileHeader.linktype, packet->pktlen);
+    LOG("enter %p %u %d", packet, pcapFileHeader.dlt, packet->pktlen);
     moloch_print_hex_string(packet->pkt, packet->pktlen);
 #endif
 
-    switch(pcapFileHeader.linktype) {
-    case 0: // NULL
+    switch(pcapFileHeader.dlt) {
+    case DLT_NULL: // NULL
         if (packet->pktlen > 4) {
             if (packet->pkt[0] == 30)
                 rc = moloch_packet_ip6(batch, packet, packet->pkt+4, packet->pktlen-4);
@@ -1207,36 +1208,41 @@ void moloch_packet_batch(MolochPacketBatch_t * batch, MolochPacket_t * const pac
             rc = MOLOCH_PACKET_CORRUPT;
         }
         break;
-    case 1: // Ether
+    case DLT_EN10MB: // Ether
         rc = moloch_packet_ether(batch, packet, packet->pkt, packet->pktlen);
         break;
-    case 12: // LOOP
-    case 101: // RAW
-        rc = moloch_packet_ip4(batch, packet, packet->pkt, packet->pktlen);
+    case DLT_RAW: // RAW
+        if ((packet->pkt[0] & 0xF0) == 0x60)
+            rc = moloch_packet_ip6(batch, packet, packet->pkt, packet->pktlen);
+        else
+            rc = moloch_packet_ip4(batch, packet, packet->pkt, packet->pktlen);
         break;
-    case 107: // Frame Relay
+    case DLT_FRELAY: // Frame Relay
         rc = moloch_packet_frame_relay(batch, packet, packet->pkt, packet->pktlen);
         break;
-    case 113: // SLL
+    case DLT_LINUX_SLL: // SLL
         if (packet->pkt[0] == 0 && packet->pkt[1] <= 4)
             rc = moloch_packet_sll(batch, packet, packet->pkt, packet->pktlen);
         else
             rc = moloch_packet_ip4(batch, packet, packet->pkt, packet->pktlen);
         break;
-    case 127: // radiotap
+    case DLT_IEEE802_11_RADIO: // radiotap
         rc = moloch_packet_radiotap(batch, packet, packet->pkt, packet->pktlen);
         break;
-    case 228: //RAW IPv4
+    case DLT_IPV4: //RAW IPv4
         rc = moloch_packet_ip4(batch, packet, packet->pkt, packet->pktlen);
         break;
-    case 239: // NFLOG
+    case DLT_IPV6: //RAW IPv4
+        rc = moloch_packet_ip6(batch, packet, packet->pkt, packet->pktlen);
+        break;
+    case DLT_NFLOG: // NFLOG
         rc = moloch_packet_nflog(batch, packet, packet->pkt, packet->pktlen);
         break;
     default:
         if (config.ignoreErrors)
             rc = MOLOCH_PACKET_CORRUPT;
         else
-            LOGEXIT("ERROR - Unsupported pcap link type %u", pcapFileHeader.linktype);
+            LOGEXIT("ERROR - Unsupported pcap link type %u", pcapFileHeader.dlt);
     }
 
     if (likely(rc == MOLOCH_PACKET_DO_PROCESS) && unlikely(packet->mProtocol == 0)) {
@@ -1631,11 +1637,44 @@ void moloch_packet_add_packet_ip(char *ipstr, int mode)
     node->data = (void *)(long)mode;
 }
 /******************************************************************************/
-void moloch_packet_set_linksnap(int linktype, int snaplen)
+void moloch_packet_set_dltsnap(int dlt, int snaplen)
 {
-    pcapFileHeader.linktype = linktype;
+    pcapFileHeader.dlt = dlt;
     pcapFileHeader.snaplen = snaplen;
     moloch_rules_recompile();
+}
+/******************************************************************************/
+void moloch_packet_set_linksnap(int linktype, int snaplen)
+{
+    // In theory you might need to do some mapping here, but we don't
+    moloch_packet_set_dltsnap(linktype, snaplen);
+}
+/******************************************************************************/
+// PCAP Header needs linktype when written
+// Code based on https://github.com/aol/moloch/issues/1303#issuecomment-554684749
+// Values from libpcap pcap-common.c
+uint32_t moloch_packet_dlt_to_linktype(int dlt)
+{
+    if (dlt <= 10 || dlt >= 104)
+        return dlt;
+
+    switch (dlt)
+    {
+#ifdef DLT_FR
+    case DLT_FR: return 107; // LINKTYPE_FRELAY;
+#endif
+    case DLT_ATM_RFC1483: return 100; // LINKTYPE_ATM_RFC1483;
+    case DLT_RAW: return 101; // LINKTYPE_RAW;
+    case DLT_SLIP_BSDOS: return 102; // LINKTYPE_SLIP_BSDOS;
+    case DLT_PPP_BSDOS: return 103; // LINKTYPE_PPP_BSDOS;
+    case DLT_C_HDLC: return 104; // LINKTYPE_C_HDLC;
+    case DLT_ATM_CLIP: return 106; // LINKTYPE_ATM_CLIP;
+    case DLT_PPP_SERIAL: return 50; // LINKTYPE_PPP_HDLC
+    case DLT_PPP_ETHER: return 51; // LINKTYPE_PPP_ETHER;
+    case DLT_PFSYNC: return 246; // LINKTYPE_PFSYNC;
+    case DLT_PKTAP: return 258; // LINKTYPE_PKTAP
+    }
+    return dlt;
 }
 /******************************************************************************/
 void moloch_packet_drophash_add(MolochSession_t *session, int which, int min)

--- a/capture/packet.c
+++ b/capture/packet.c
@@ -1232,7 +1232,7 @@ void moloch_packet_batch(MolochPacketBatch_t * batch, MolochPacket_t * const pac
     case DLT_IPV4: //RAW IPv4
         rc = moloch_packet_ip4(batch, packet, packet->pkt, packet->pktlen);
         break;
-    case DLT_IPV6: //RAW IPv4
+    case DLT_IPV6: //RAW IPv6
         rc = moloch_packet_ip6(batch, packet, packet->pkt, packet->pktlen);
         break;
     case DLT_NFLOG: // NFLOG

--- a/capture/plugins/daq/reader-daq.c
+++ b/capture/plugins/daq/reader-daq.c
@@ -91,7 +91,7 @@ void reader_daq_start() {
     int err;
 
     //ALW - Bug: assumes all linktypes are the same
-    moloch_packet_set_linksnap(daq_get_datalink_type(module, handles[0]), config.snapLen);
+    moloch_packet_set_dltsnap(daq_get_datalink_type(module, handles[0]), config.snapLen);
 
     int i;
     for (i = 0; i < MAX_INTERFACES && config.interface[i]; i++) {

--- a/capture/plugins/pfring/reader-pfring.c
+++ b/capture/plugins/pfring/reader-pfring.c
@@ -88,8 +88,6 @@ LOCAL void *reader_pfring_thread(void *posv)
 }
 /******************************************************************************/
 void reader_pfring_start() {
-    int dlt_to_linktype(int dlt);
-
     moloch_packet_set_dltsnap(DLT_EN10MB, config.snapLen);
 
     int i;

--- a/capture/plugins/pfring/reader-pfring.c
+++ b/capture/plugins/pfring/reader-pfring.c
@@ -90,7 +90,7 @@ LOCAL void *reader_pfring_thread(void *posv)
 void reader_pfring_start() {
     int dlt_to_linktype(int dlt);
 
-    moloch_packet_set_linksnap(1, config.snapLen);
+    moloch_packet_set_dltsnap(DLT_EN10MB, config.snapLen);
 
     int i;
     for (i = 0; i < MAX_INTERFACES && config.interface[i]; i++) {

--- a/capture/plugins/snf/reader-snf.c
+++ b/capture/plugins/snf/reader-snf.c
@@ -121,7 +121,7 @@ LOCAL void *reader_snf_thread(gpointer posv)
 }
 /******************************************************************************/
 void reader_snf_start() {
-    moloch_packet_set_linksnap(DLT_EN10MB, config.snapLen);
+    moloch_packet_set_dltsnap(DLT_EN10MB, config.snapLen);
 
     int ringStartOffset = (snfProcNum-1)*snfNumRings;
     int i, r;

--- a/capture/plugins/writer-s3.c
+++ b/capture/plugins/writer-s3.c
@@ -627,7 +627,9 @@ void writer_s3_create(const MolochPacket_t *packet)
 
     outputBuffer = moloch_http_get_buffer(config.pcapWriteSize + MOLOCH_PACKET_MAX_LEN);
     outputPos = 0;
-    append_to_output(&pcapFileHeader, 24, FALSE, 0);
+    uint32_t linktype = moloch_packet_dlt_to_linktype(pcapFileHeader.dlt);
+    append_to_output(&pcapFileHeader, 20, FALSE, 0);
+    append_to_output(&linktype, 4, FALSE, 0);
     make_new_block();                   // So we can read the header in a small amount of data fetched
 
     if (config.debug)

--- a/capture/reader-libpcap-file.c
+++ b/capture/reader-libpcap-file.c
@@ -557,7 +557,6 @@ LOCAL gboolean reader_libpcapfile_read()
 /******************************************************************************/
 LOCAL void reader_libpcapfile_opened()
 {
-    int dlt_to_linktype(int dlt);
     int moloch_db_can_quit();
 
     if (config.flushBetween) {

--- a/capture/reader-libpcap-file.c
+++ b/capture/reader-libpcap-file.c
@@ -574,11 +574,11 @@ LOCAL void reader_libpcapfile_opened()
         }
     }
 
-    moloch_packet_set_linksnap(dlt_to_linktype(pcap_datalink(pcap)) | pcap_datalink_ext(pcap), pcap_snapshot(pcap));
+    moloch_packet_set_dltsnap(pcap_datalink(pcap), pcap_snapshot(pcap));
 
     offlineFile = pcap_file(pcap);
 
-    if (config.bpf && pcapFileHeader.linktype != 239) {
+    if (config.bpf && pcapFileHeader.dlt != DLT_NFLOG) {
         struct bpf_program   bpf;
 
         if (pcap_compile(pcap, &bpf, config.bpf, 1, PCAP_NETMASK_UNKNOWN) == -1) {

--- a/capture/reader-libpcap.c
+++ b/capture/reader-libpcap.c
@@ -91,8 +91,6 @@ LOCAL void *reader_libpcap_thread(gpointer posv)
 }
 /******************************************************************************/
 void reader_libpcap_start() {
-    int dlt_to_linktype(int dlt);
-
     //ALW - Bug: assumes all linktypes are the same
     moloch_packet_set_dltsnap(pcap_datalink(pcaps[0]), pcap_snapshot(pcaps[0]));
 

--- a/capture/reader-libpcap.c
+++ b/capture/reader-libpcap.c
@@ -94,7 +94,7 @@ void reader_libpcap_start() {
     int dlt_to_linktype(int dlt);
 
     //ALW - Bug: assumes all linktypes are the same
-    moloch_packet_set_linksnap(dlt_to_linktype(pcap_datalink(pcaps[0])) | pcap_datalink_ext(pcaps[0]), pcap_snapshot(pcaps[0]));
+    moloch_packet_set_dltsnap(pcap_datalink(pcaps[0]), pcap_snapshot(pcaps[0]));
 
     int i;
     for (i = 0; i < MAX_INTERFACES && config.interface[i]; i++) {

--- a/capture/reader-tpacketv3.c
+++ b/capture/reader-tpacketv3.c
@@ -203,7 +203,7 @@ void reader_tpacketv3_init(char *UNUSED(name))
         LOGEXIT("block size %d not divisible by %u", blocksize, config.snapLen);
     }
 
-    moloch_packet_set_linksnap(1, config.snapLen);
+    moloch_packet_set_dltsnap(DLT_EN10MB, config.snapLen);
 
     pcap_t *dpcap = pcap_open_dead(pcapFileHeader.linktype, pcapFileHeader.snaplen);
 

--- a/capture/reader-tpacketv3.c
+++ b/capture/reader-tpacketv3.c
@@ -205,7 +205,7 @@ void reader_tpacketv3_init(char *UNUSED(name))
 
     moloch_packet_set_dltsnap(DLT_EN10MB, config.snapLen);
 
-    pcap_t *dpcap = pcap_open_dead(pcapFileHeader.linktype, pcapFileHeader.snaplen);
+    pcap_t *dpcap = pcap_open_dead(pcapFileHeader.dlt, pcapFileHeader.snaplen);
 
     if (config.bpf) {
         if (pcap_compile(dpcap, &bpf, config.bpf, 1, PCAP_NETMASK_UNKNOWN) == -1) {

--- a/capture/rules.c
+++ b/capture/rules.c
@@ -713,7 +713,7 @@ void moloch_rules_recompile()
     if (deadPcap)
         pcap_close(deadPcap);
 
-    deadPcap = pcap_open_dead(pcapFileHeader.linktype, pcapFileHeader.snaplen);
+    deadPcap = pcap_open_dead(pcapFileHeader.dlt, pcapFileHeader.snaplen);
     MolochRule_t *rule;
     for (t = 0; t < MOLOCH_RULE_TYPE_NUM; t++) {
         for (r = 0; (rule = current.rules[t][r]); r++) {
@@ -721,7 +721,7 @@ void moloch_rules_recompile()
                 continue;
 
             pcap_freecode(&rule->bpfp);
-            if (pcapFileHeader.linktype != 239) {
+            if (pcapFileHeader.dlt != DLT_NFLOG) {
                 if (pcap_compile(deadPcap, &rule->bpfp, rule->bpf, 1, PCAP_NETMASK_UNKNOWN) == -1) {
                     LOGEXIT("ERROR - Couldn't compile filter %s: '%s' with %s", rule->filename, rule->bpf, pcap_geterr(deadPcap));
                 }

--- a/capture/writer-disk.c
+++ b/capture/writer-disk.c
@@ -340,7 +340,9 @@ LOCAL void writer_disk_create(MolochPacket_t * const packet)
     clock_gettime(CLOCK_REALTIME_COARSE, &outputFileTime);
 
     output->fileId = outputId;
-    memcpy(output->buf, &pcapFileHeader, 24);
+    memcpy(output->buf, &pcapFileHeader, 20);
+    uint32_t linktype = moloch_packet_dlt_to_linktype(pcapFileHeader.dlt);
+    memcpy(output->buf+20, &linktype, 4);
 }
 /******************************************************************************/
 struct pcap_timeval {

--- a/capture/writer-simple.c
+++ b/capture/writer-simple.c
@@ -266,7 +266,7 @@ LOCAL char *writer_simple_get_kekId ()
         {
             int namelen = strlen(config.nodeName);
             int bufboundary = j + namelen;
-            
+
             if(bufboundary >= (int) sizeof(okek)) {
                 LOGEXIT("ERROR - node name '%s' is too long", config.nodeName);
             }
@@ -351,7 +351,10 @@ LOCAL void writer_simple_write(const MolochSession_t * const session, MolochPack
             LOGEXIT("ERROR - pcap open failed - Couldn't open file: '%s' with %s  (%d) -- You may need to check directory permissions or set pcapWriteMethod=simple-nodirect in config.ini file.  See https://molo.ch/settings#pcapwritemethod", name, strerror(errno), errno);
         }
         info->file->pos = currentInfo[thread]->bufpos = 24;
-        memcpy(info->buf, &pcapFileHeader, 24);
+
+        memcpy(info->buf, &pcapFileHeader, 20);
+        uint32_t linktype = moloch_packet_dlt_to_linktype(pcapFileHeader.dlt);
+        memcpy(info->buf+20, &linktype, 4);
         if (config.debug)
             LOG("opened %d %s %d", thread, name, info->file->fd);
         g_free(name);

